File: results/0.0/ARP/arp-gpt-4-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// Define ARP packet structure
typedef struct {
    uint16_t htype;
    uint16_t ptype;
    uint8_t hlen;
    uint8_t plen;
    uint16_t oper;
    uint8_t sha[6];
    uint8_t spa[4];
    uint8_t tha[6];
    uint8_t tpa[4];
} ARP_Packet;

// Function to parse ARP packet
HParser *create_arp_parser() {
    HParser *htype = h_uint16();
    HParser *ptype = h_uint16();
    HParser *hlen = h_uint8();
    HParser *plen = h_uint8();
    HParser *oper = h_uint16();
    HParser *sha = h_repeat_n(h_uint8(), 6);
    HParser *spa = h_repeat_n(h_uint8(), 4);
    HParser *tha = h_repeat_n(h_uint8(), 6);
    HParser *tpa = h_repeat_n(h_uint8(), 4);

    return h_sequence(htype, ptype, hlen, plen, oper, sha, spa, tha, tpa, NULL);
}

// Function to print ARP packet
void print_arp_packet(const ARP_Packet *packet) {
    printf("Hardware Type: %u\n", packet->htype);
    printf("Protocol Type: %u\n", packet->ptype);
    printf("Hardware Length: %u\n", packet->hlen);
    printf("Protocol Length: %u\n", packet->plen);
    printf("Operation: %u\n", packet->oper);
    printf("Sender Hardware Address: ");
    for (int i = 0; i < 6; i++) {
        printf("%02X", packet->sha[i]);
        if (i < 5) printf(":");
    }
    printf("\nSender Protocol Address: ");
    for (int i = 0; i < 4; i++) {
        printf("%d", packet->spa[i]);
        if (i < 3) printf(".");
    }
    printf("\nTarget Hardware Address: ");
    for (int i = 0; i < 6; i++) {
        printf("%02X", packet->tha[i]);
        if (i < 5) printf(":");
    }
    printf("\nTarget Protocol Address: ");
    for (int i = 0; i < 4; i++) {
        printf("%d", packet->tpa[i]);
        if (i < 3) printf(".");
    }
    printf("\n");
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <binary file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *buffer = malloc(file_size);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    fread(buffer, 1, file_size, file);
    fclose(file);

    HParser *arp_parser = create_arp_parser();
    HParseResult *result = h_parse(arp_parser, buffer, file_size);
    if (result && result->ast) {
        ARP_Packet *packet = (ARP_Packet *)result->ast->seq->elements[0]->user;
        print_arp_packet(packet);
        h_parse_result_free(result);
    } else {
        fprintf(stderr, "Failed to parse ARP packet\n");
    }

    free(buffer);
    h_parser_free(arp_parser);
    return 0;
}

File: results/0.0/ARP/arp-gpt-4-turbo.rs
use nom::{
    bytes::complete::take,
    number::complete::{be_u16, be_u8},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::{self, Read};

#[derive(Debug)]
struct ARPHeader {
    htype: u16,
    ptype: u16,
    hlen: u8,
    plen: u8,
    oper: u16,
    sha: Vec<u8>,
    spa: Vec<u8>,
    tha: Vec<u8>,
    tpa: Vec<u8>,
}

fn parse_arp_header(input: &[u8]) -> IResult<&[u8], ARPHeader> {
    let (input, htype) = be_u16(input)?;
    let (input, ptype) = be_u16(input)?;
    let (input, hlen) = be_u8(input)?;
    let (input, plen) = be_u8(input)?;
    let (input, oper) = be_u16(input)?;
    let (input, sha) = take(hlen as usize)(input)?;
    let (input, spa) = take(plen as usize)(input)?;
    let (input, tha) = take(hlen as usize)(input)?;
    let (input, tpa) = take(plen as usize)(input)?;

    Ok((
        input,
        ARPHeader {
            htype,
            ptype,
            hlen,
            plen,
            oper,
            sha: sha.to_vec(),
            spa: spa.to_vec(),
            tha: tha.to_vec(),
            tpa: tpa.to_vec(),
        },
    ))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_arp_header(&buffer) {
        Ok((_, arp_header)) => {
            println!("{:?}", arp_header);
        }
        Err(e) => {
            eprintln!("Failed to parse ARP header: {:?}", e);
        }
    }

    Ok(())
}